pretyping/recordops.ml

```diff
@@ -298,29 +298,40 @@ let add_canonical_structure x = Lib.add_anonymous_leaf (inCanonStruc x)
 
 (*s High-level declaration of a canonical structure *)
 
-let error_not_structure ref =
+let error_not_structure ref description =
   user_err ~hdr:"object_declare"
-    (Id.print (basename_of_global ref) ++ str" is not a structure object.")
+    (str"Could not declare a canonical structure " ++
+       (Id.print (basename_of_global ref) ++ str"." ++ spc() ++
+          str(description)))
 
 let check_and_decompose_canonical_structure ref =
-  let sp = match ref with ConstRef sp -> sp | _ -> error_not_structure ref in
+  let sp =
+    match ref with
+      ConstRef sp -> sp
+    |  _ -> error_not_structure ref "Expected an instance of a record or structure."
+  in
   let env = Global.env () in
   let u = Univ.make_abstract_instance (Environ.constant_context env sp) in
   let vc = match Environ.constant_opt_value_in env (sp, u) with
     | Some vc -> vc
-    | None -> error_not_structure ref in
+    | None -> error_not_structure ref "Could not find its value in the global environment." in
```

Comment:
This error message has no corresponding test in either this PR or #6703.

Reply:
I could not find ways to trigger the two errors that you've highlighted. I can try again, or maybe someone knows how they can happen.